### PR TITLE
WebSocket refactoring, bug fixes, improvements, and passes most Autobahn Test (The most standard WebSocket Protocol Test).

### PR DESCRIPTION
**Pull request by dumganhar**
This PR is based on the latest release version of `libwebsockets`, so before merging it, please upgrade `libwebsockets`
## Result

It has passed most of Autobahn Test (The most standard WebSocket Protocol Test).
Please see the result at http://182.254.241.97/runtime-test/websocket-report/index.html
### Refactoring
- Adds more callback functions like `onConnectionOpened`, `onConnectionClosed` rather than put all websocket callback in just one `onSocketCallback`, code will be clearer and readable.

``` c++
    int onSocketCallback(struct lws *wsi, int reason, void *user, void *in, ssize_t len);

    void onClientWritable();
    void onClientReceivedData(void* in, ssize_t len);
    void onConnectionOpened();
    void onConnectionError();
    void onConnectionClosed();
```
- Uses `performFunctionInCocosThread(cb);` to send msg to Cocos Thread rather than make a scheduler update
- Uses `LOGD` for outputting log to debugging websocket connections since websocket stuff uses a sub thread. CCLOG isn't threadsafe.
### Bug Fixes
- WebSocket::close should not block Cocos Thread, it should just notify websocket thread to exit. Now the behavior is the same as Browser. To prove, you could type some test code in Chrome and the result will be:

``` javascript
ws = new WebSocket("ws://127.0.0.1:9123"); ws.onclose = function(){ console.log("I'm closed");}; 
(){ console.log("I'm closed");}
console.log("before close"); ws.close(); console.log("hello");
VM437:2 before close
VM437:2 hello
undefined
VM433:2 I'm closed
```
- Fixed every websocket event will cost 50ms at least, since the old way was wrong, we should not wait 50ms in sub thread while loop. `lws_service(xxx, time_to_wait)` will do it for us. So it fixed delay of message receive/send.
- Fixed potential high CPU cost since the old code will trigger a websocket writable event every time after the writable callback. Now only when the size of the sending message queue isn't zero, a `lws_callback_on_writable` will be invoked
- Fixed `libwebsocket_context_destroy` may not be invoked while closing connection.
- Fixed sending large data will trigger websocket errors
- Fixed receiving large data in many frames will trigger JS `onmessage` callback many times
- Fixed `lws_write` may return a smaller value than the value `n` for sending, a `WebSocketFrame` class is doing the job. `data->issued` should not be increased while a `WebSocketFrame` isn't totally sent.
- Fixed connection isn't closed if received text message is an invalid UTF8 string.
- Fixed the memory leak that JSB_WebSocketDelegate is never released.
### Improvements
- Better performance for received data buffer. Uses `std::vector` for dynamic received data buffer, the old way is every time there is a frame received, buffer memory will be extended every time there is a message comes. It was very low performance.
- Supports receive/send data with zero size.
- Larger socket rx/tx buffer size `WS_RX_BUFFER_SIZE` for better performance.
- Adds static method `WebSocket::closeAllConnections`, it should be invoked while resetting `Director`, for ensure all websocket threads exit. Otherwise, cocos will release all engine classes when exiting game, then if a websocket event is coming, it will cause exception since websocket API is asynchronous.

### CODE REVIEW

File: cocos/network/WebSocket.cpp [hunk 1]
```diff
@@ -120,27 +170,22 @@ WsThreadHelper::WsThreadHelper()
...
 }
 
 bool WsThreadHelper::createThread(const WebSocket& ws)
 {
     _ws = const_cast<WebSocket*>(&ws);
-    
+
     // Creates websocket thread
-    _subThreadInstance = new (std::nothrow) std::thread(&WsThreadHelper::wsThreadEntryFunc, this);
+    _subThreadInstance = new std::thread(&WsThreadHelper::wsThreadEntryFunc, this);
```
Comment (pandamicro) [inline cocos/network/WebSocket.cpp]: Why remove nothrow declaration ?

File: cocos/network/WebSocket.cpp [hunk 2]
```diff
@@ -151,22 +196,22 @@ void WsThreadHelper::quitSubThread()
...
+        _ws->onSubThreadLoop();
     }
+
+    _ws->onSubThreadEnded();
     
+    LOGD("WebSocket thread exit, helper instance: %p\n", this);
 }
 
-void WsThreadHelper::sendMessageToUIThread(WsMessage *msg)
+void WsThreadHelper::sendMessageToUIThread(const std::function<void()>& cb)
```
Comment (pandamicro) [inline cocos/network/WebSocket.cpp]: This function name seems no longer suitable

File: cocos/network/WebSocket.cpp [hunk 3]
```diff
@@ -241,21 +323,42 @@ WebSocket::WebSocket()
...
+    if (__websocketInstances != nullptr)
+    {
+        auto iter = std::find(__websocketInstances->begin(), __websocketInstances->end(), this);
+        if (iter != __websocketInstances->end())
+        {
+            __websocketInstances->erase(iter);
+        }
+        else
+        {
+            LOGD("ERROR: WebSocket instance (%p) added to container!\n", this);
```
Comment (pandamicro) [inline cocos/network/WebSocket.cpp]: To be more clear
ERROR: WebSocket instance (%p) haven't been added to global list!

File: cocos/network/WebSocket.cpp [hunk 4]
```diff
@@ -311,33 +414,35 @@ bool WebSocket::init(const Delegate& delegate,
     {
         protocolCount = 1;
     }
-    
-	_wsProtocols = new (std::nothrow) libwebsocket_protocols[protocolCount+1];
-	memset(_wsProtocols, 0, sizeof(libwebsocket_protocols)*(protocolCount+1));
+
+    _wsProtocols = new lws_protocols[protocolCount+1];
```
Comment (pandamicro) [inline cocos/network/WebSocket.cpp]: Normally, we should always add `nothrow` to new statement, riq and Wenhai have cleaned up our codes, if there is no special reason, we shouldn't remove it.

File: cocos/network/WebSocket.cpp [hunk 5]
```diff
@@ -311,33 +414,35 @@ bool WebSocket::init(const Delegate& delegate,
...
+    memset(_wsProtocols, 0, sizeof(lws_protocols)*(protocolCount+1));
 
     if (protocols && protocols->size() > 0)
     {
         int i = 0;
         for (std::vector<std::string>::const_iterator iter = protocols->begin(); iter != protocols->end(); ++iter, ++i)
         {
-            char* name = new (std::nothrow) char[(*iter).length()+1];
+            char* name = new char[(*iter).length()+1];
```
Comment (pandamicro) [inline cocos/network/WebSocket.cpp]: Same as above

**Comment by pandamicro** [comment]
Thanks @dumganhar Great PR !

The logic is much more clear with better stability and performance. I have no further questions except some minor remark in the previous comments. However, I'm not very familiar with network behavior or websocket, and I'm not the owner of network module. @linshun Can you assign someone to give a final confirm on this PR ?

**Comment by dumganhar** [comment]
@pandamicro ,all you commented were fixed.

**Comment by linshun** [comment]
Hi @ricardoquesada, you are the owner of networking and could you please review this PR and merge it? Thanks.

It is a great improvement for WebSocket.

**Comment by ricardoquesada** [comment]
looks good to me.

@dumganhar what is the use case for cocos2d-x ? Is it only valid for cocos2d-js users ?
BTW, the build is red. If you fix it, I'll merge it.

**Comment by dumganhar** [comment]
@ricardoquesada , i'll update libwebsocket tomorrow.

**Comment by dumganhar** [comment]
Updates https://github.com/cocos2d/cocos2d-x-3rd-party-libs-src/pull/78
Could somebody build the libwebsockets library for all platforms?

**Comment by dumganhar** [comment]
@xiaofeng11 , may i ask when will this Pull Request be merged?

**Comment by dumganhar** [comment]
@xiaofeng11 , may i ask when will this Pull Request be merged?
+1

**Comment by dumganhar** [comment]
> @dumganhar what is the use case for cocos2d-x ? Is it only valid for cocos2d-js users ?
> BTW, the build is red. If you fix it, I'll merge it.

@ricardoquesada , my changes are also valid for C++ & Lua users.
WebSocket.h/.cpp is a wrapper for libwebsockets, and could be used for C++/Lua/JS users.

**Comment by deviantmk** [comment]
bumb

**Comment by ricardoquesada** [comment]
@dumganhar the build is still red. could you fix the issues that prevent it from building ok?

**Comment by dumganhar** [comment]
@ricardoquesada, 
@xiaofeng11 has been fixing the `red` problem. I don't know the process, so i am asking the status of updating the 3rd `libwebsockets` here. But got no response. :(

**Comment by pandemosth** [comment]
@dumganhar could you have a look at my question on https://github.com/cocos2d/cocos2d-x-3rd-party-libs-src/pull/78 regarding why SSL support has been reverted in your change?
Thanks

**Comment by xiaofeng11** [comment]
after pr https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin/pull/199 been merged, change deps lib ref in this PR and pull my branch https://github.com/xiaofeng11/cocos2d-x/tree/v3_libwebsockets to combine the change of cocos/2d/win10_props/cocos2d_win10_app.props and cocos/2d/winrt_8.1_props/cocos2d_winrt_8.1_app.props to get pass of jenkins auto build

**Comment by dumganhar** [comment]
Hi @pandemosth ,
Don't worry, I think @xiaofeng11 has compiled it with SSL support yet.

**Comment by xiaofeng11** [comment]
@dumganhar The libwebsockets update is compile by cocos2d-x-3rd-party-libs-src compile scripts, it won't have ssl support

**Comment by pandemosth** [comment]
Can that be fixed? Websocket client really should support ssl.

**Comment by dumganhar** [comment]
The previous release version of Cocos2dx also doesn't support WebSocket with SSL, right?
Anyway, whether it was supported or not, the changes of this PR isn't relevant to it.
@xiaofeng11 , could you compile `libwebsockets` with SSL support?

**Comment by xiaofeng11** [comment]
@pandamicro @dumganhar Sorry, my mistake. New libwebsocket is compiled with openSSL support.
But I can not sure if Win10 & Win8.1 have same feature, because these lib can not compile with scripts. So I create project file from win32 src and modify code as patch in MS fork to compile these 2 lib

**Comment by zilongshanren** [comment]
@dumganhar 
The 3rd party bin has been merged, please check and update the config.json file.
Thanks.

**Comment by dumganhar** [comment]
@zilongshanren ,ok, i will.

**Comment by dumganhar** [comment]
Update config.json DONE.

**Comment by xiaofeng11** [comment]
@dumganhar as mentioned above, you need pull my branch at https://github.com/xiaofeng11/cocos2d-x/tree/v3_libwebsockets to merge change to cocos/2d/win10_props/cocos2d_win10_app.props and cocos/2d/winrt_8.1_props/cocos2d_winrt_8.1_app.props to get this pr pass jinkens auto build.

**Comment by dumganhar** [comment]
@xiaofeng11 ,ok , i will check it.

**Comment by dumganhar** [comment]
@xiaofeng11 , DONE for merging  https://github.com/xiaofeng11/cocos2d-x/tree/v3_libwebsockets

**Comment by dumganhar** [comment]
@xiaofeng11 , build passed. Please check.

**Comment by xiaofeng11** [comment]
@pandamicro @zilongshanren I think this pr is OK for merge.

**Comment by dumganhar** [comment]
hey, @all, how about this PR now?

**Comment by dumganhar** [comment]
Hello, could this Pull Request be merged now ?

**Comment by zilongshanren** [comment]
@dumganhar 
I'm fine with this PR, @ricardoquesada  thoughts?

**Comment by dumganhar** [comment]
It's really disappointed that there isn't any updates. :(

**Comment by ricardoquesada** [comment]
good to me. merged

**Comment by dumganhar** [comment]
@ricardoquesada, thank you. :)

**Comment by stammen** [comment]
@ricardoquesada @dumganhar This pull request broke the Windows Universal 8.1 ARM builds. What version of libwebsockets did you use? I will compile all of the libs and submit a pull request to cocos2d-x-3rd-party-libs-bin. Should I add OpenSSL support?

**Comment by zilongshanren** [comment]
@stammen  maybe @xiaofeng11  knows this issue better, he made all the prebuilt libs for windows universal 8.1.

**Comment by xiaofeng11** [comment]
@stammen This pr use 15c92b1 commit on master branch, which commits on Jan 18, 2016. You can use https://github.com/cocos2d/cocos2d-x-3rd-party-libs-src scripts to get this version.

And could you please build a win32 version either? I have compiled a win32 version, but my version can not compatible with Visual Studio 2015, compile cpp-tests with my version libwebsockets.lib, I get __imp____iob_func & __imp___vsnprintf missing error at link. But it work fine with Visual Studio 2013.

**Comment by dannyjiajia** [comment]
hi guys, why disable the `onError` callback in Lua ? https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/scripting/lua-bindings/manual/network/Lua_web_socket.cpp#L136. @dumganhar

**Comment by roed** [comment]
When I try the new websockets code, my win32 machine doesn't connect to my node.js backend anymore. It always ends up with ErrorCode::CONNECTION_FAILURE.

**Comment by dumganhar** [comment]
@dannyjiajia , Sorry for my late reply, I don't know why it was commented. 
And this pull request didn't comment that line. Anyway, I will find out why it was commented.

**Comment by dumganhar** [comment]
@roed , how to reproduce that problem? Could you send me a node.js demo for reproduce? Thanks.

**Comment by dannyjiajia** [comment]
@dumganhar thanks, any update?

**Comment by dumganhar** [comment]
@dannyjiajia, I checked the git commit tree. Found it was commented at the beginning of v3.x.
Yep, it should not be commented. I created an issue (https://github.com/cocos2d/cocos2d-x/issues/15904) and will try to fix it.

**Comment by dannyjiajia** [comment]
@dumganhar thanks :)

**Comment by dumganhar** [comment]
@dannyjiajia , the issue was fixed at https://github.com/cocos2d/cocos2d-x/pull/15922
Thanks for your feedback.

**Comment by roed** [comment]
@dumganhar I ran the cocos2d-x test case and it did work with a socket.io test server. In my project, I only replaced some libraries and the websocket code, which didn't work. Will do a full upgrade to 3.11.1 and see if the code still breaks. So you can ignore my question, sorry :)

**Comment by dumganhar** [comment]
@roed, OK. Feel free of reporting anything after you upgrade to 3.11.1. Thanks.

**Comment by roed** [comment]
@dumganhar I upgraded to 3.11.1, win32 works fine but now I have an issue on iOS. My socket.io server code is nothing more than something like:

```
var app = require('http').createServer();
var io = require('socket.io').listen(app);
app.listen(3000);
io.on('connection', function (socket) {
    console.log('Socket connected: ', socket.id);
});
```

Somehow lws_client_connect in websockets.cpp returns a nullptr and doesn't give me anything else in the logging output. Do you have any idea what this could be? I also tried compiling the library myself, but with the same issues.

**Pull request merged**